src/librvnpal/src/posix/fileutils.c

```diff
@@ -0,0 +1,191 @@
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <sys/mman.h>
+#include <assert.h>
+#include <time.h>
+
+#include "rvn.h"
+#include "status_codes.h"
+#include "internal_posix.h"
+
+PRIVATE int64_t
+_nearest_size_to_page_size(int64_t orig_size, int64_t sys_page_size)
+{
+    int64_t mod = orig_size % sys_page_size;
+    if (mod == 0)
+    {
+        return orig_size;
+    }
+    return ((orig_size / sys_page_size) + 1) * sys_page_size;
+}
+
+PRIVATE int64_t
+_pwrite(int32_t fd, void *buffer, uint64_t count, uint64_t offset, int32_t *detailed_error_code)
+{
+    uint64_t actually_written = 0;
+    int64_t cifs_retries = 3;
+    do
+    {
+        int64_t result = pwrite(fd, buffer, count - actually_written, offset + actually_written);
+        if (result < 0) /* we assume zero cannot be returned at any case as defined in POSIX */
+        {
+            if (errno == EINVAL && _sync_directory_allowed(fd) == SYNC_DIR_NOT_ALLOWED && --cifs_retries > 0)
+            {
+                /* cifs/nfs mount can sometimes fail on EINVAL after file creation
+                lets give it few retries with short pauses between them - RavenDB-11954 */
+                struct timespec ts;
+                ts.tv_sec = 0;
+                ts.tv_nsec = 200000000L; /* 200mSec */
+                nanosleep(&ts, NULL);
+                continue; /* retry cifs */
+            }
+            *detailed_error_code = errno;
+            if (cifs_retries != 3)
+                return FAIL_PWRITE_WITH_RETRIES;
+            return FAIL_PWRITE;
+        }
+        actually_written += result;
+    } while (actually_written < (int64_t)count);
+
+    return SUCCESS;
+}
+
+PRIVATE int32_t
+_allocate_file_space(int32_t fd, int64_t size, int32_t *detailed_error_code)
+{
+    int32_t result;
+    int32_t retries;
+    for (retries = 0; retries < 1024; retries++)
+    {
+        result = _rvn_fallocate(fd, 0, size);
+
+        switch (result)
+        {
+        case EINVAL:
+        case EFBIG: /* can occure on >4GB allocation on fs such as ntfs-3g, W95 FAT32, etc.*/
+            /* fallocate is not supported, we'll use lseek instead */
+            {
+                char b = 0;
+                int64_t rc = _pwrite(fd, &b, 1UL, (uint64_t)size - 1UL, detailed_error_code);
+                if (rc != SUCCESS)
+                    *detailed_error_code = errno;
+                return rc;
+            }
+            break;
+        case EINTR:
+            *detailed_error_code = errno;
+            continue; /* retry */
+
+        case SUCCESS:
+            return SUCCESS;
+
+        default:
+            *detailed_error_code = result;
+            return FAIL_ALLOC_FILE;
+        }
+    }
+    return result; /* return EINTR */
+}
+
+PRIVATE int32_t
+_ensure_path_exists(const char* path)
+{
+    /* TODO: implement */
+    return SUCCESS;   
+}
+
+PRIVATE int32_t 
+_open_file_to_read(const char *file_name, void **handle, int32_t *detailed_error_code)
```

Comment:
In open for write, we are using `journal_handle`, here we are storing the raw fd in the void*.
We need to be consistent, because otherwise, it can get messy

Reply:
This is an internal impl detail, use it as such 
```suggestion
_open_file_to_read(const char *file_name, int *fd, int32_t *detailed_error_code)
```

Reply:
No. This is (will, when windows' map pager will be pal implemented) need to fit for both platforms. So the header must be **handle. Unless we want different headers (_open_file_to_read_linux and _open_file_to_read_windows, and call the appropriate from the posix/win rvn_* func)

Reply:
This is private for the impl. No need to expose the xplat stuff here.